### PR TITLE
#259-Generate set of mutually unbiased bases

### LINKED ISSUE #259: Feature: Generate a set of d-dimensional mutually unbiased bases (for prime d)
This feature would allow the user to generate a set of d-dimensional [mutually unbiased bases](https://en.wikipedia.org/wiki/Mutually_unbiased_bases). Note that this procedure only works when the dimension "d" is prime (otherwise, it is unknown how to generate such a set in general).

In order to generate such a set, we will require a helper function to determine if a given integer is a prime power:
```py
import numpy as np
from toqito.matrices import gen_pauli
from sympy import isprime
```

```py
def is_prime_power(n: int) -> bool:
    """Determine if number is prime power."""
    p = 2
    primes = []
    for i in range(2, n + 1):
        if isprime(i):
            primes.append(i)
    for i in primes:
        if n % i == 0:
            c = 0
            while n % i == 0:
                n //= i
                c += 1
            if n == 1:
                return True
            return False
```

Putting the above together, we can write our function:

```py
def mubs(dim: int) -> list[np.ndarray]:
    """Function to generate list of MUBs for a given dimension.
    
    Note that this function only works if the dimension provided is prime 
    or a power of a prime. Otherwise, we don't know how to generate general MUBs.
    """
    
    # The first basis will always be the standard basis:
    mats = [np.eye(dim)]

    pauli_x = gen_pauli(1, 0, dim)
    if isprime(dim):
        pauli_z = gen_pauli(0, 1, dim)

        for j in range(dim, 0, -1):
            _, eigen_vec = np.linalg.eig(pauli_x @ pauli_z**(j))
            mats.append(eigen_vec)
    elif is_prime_power(dim) and not isprime(dims):
        raise(f"Dimension {dim} is a prime power but not prime (more complicated no support atm).")
    else:
        raise(f"No general construction of MUBs is known for dimension: {dim}.")

    mubs: list[np.ndarray] = []
    for i in range(len(mats)):
        nrows, ncols = mats[i].shape[0], mats[i].shape[1]
        for row in range(nrows):
            mub: list[np.ndarray] = []
            for col in range(ncols):
                mub.append(mats[i][col][row])
            mubs.append(np.array(mub))
    return mubs
```

While most of the above does the heavy lifting, we still require unit tests and proper documentation.

### PR DESCRIPTION
**Pull request by vprusso**
Closes #259 

- Feature to generate a MUB in prime dimensions
- Improving function to check if set of vectors is a MUB
- Tests for both MUB and check if MUB.

### CODE REVIEW

File: toqito/state_props/is_mutually_unbiased_basis.py [hunk 1]
```diff
@@ -80,17 +78,25 @@ def is_mutually_unbiased_basis(vec_list: list[np.ndarray | list[float | Any]]) -
     :return: :code:`True` if :code:`vec_list` constitutes a mutually unbiased basis, and
              :code:`False` otherwise.
     """
-    if len(vec_list) <= 1:
+    if len(vectors) <= 1:
         raise ValueError("There must be at least two bases provided as input.")
```
Comment (purva-thakre) [inline toqito/state_props/is_mutually_unbiased_basis.py:82]: Should we make an issue for checking what happens when `is_check_something` has been provided an empty array as an input? 

For the current function, I think the output will be `true` if 2 or more empty arrays are provided as input. In the tests, I do see 1 function checking for the `ValueError` on line 82.

File: toqito/states/mutually_unbiased_basis.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    .. math::
+        M_0 = \left\{|0\rangle, |1\rangle \right\}, \\
+        M_1 = \left\{\frac{|0\rangle + |1\rangle}{\sqrt{2}}, \frac{|0\rangle - |1\rangle}{\sqrt{2}}\right\}
+        M_2 = \left\{\frac{|0\rangle + i|1\rangle}{\sqrt{2}}, \frac{|0\rangle - i|1\rangle}{\sqrt{2}}\right\}
+
+    References
+    ==========
+    .. [WikMUB] Wikipedia: Mutually unbiased bases
+        https://en.wikipedia.org/wiki/Mutually_unbiased_bases
+
```
Comment (purva-thakre) [inline toqito/states/mutually_unbiased_basis.py:28]: Same as what I said in `state_props/is_mutually_unbiased_basis`. 

I might be wrong but I don't think this reference has been cited in this function's docstring. So, using `docname in docnames` won't automatically generate a reference.

File: toqito/states/mutually_unbiased_basis.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    pauli_x = gen_pauli(1, 0, dim)
+    if isprime(dim):
+        pauli_z = gen_pauli(0, 1, dim)
+
+        for j in range(dim, 0, -1):
+            _, eigen_vec = np.linalg.eig(pauli_x @ pauli_z**(j))
+            mats.append(eigen_vec)
+    elif _is_prime_power(dim) and not isprime(dim):
+        raise ValueError(f"Dimension {dim} is a prime power but not prime (more complicated no support atm).")
```
Comment (purva-thakre) [inline toqito/states/mutually_unbiased_basis.py:44]:
```suggestion
        raise ValueError(f"Dimension {dim} is a prime power but not prime (more complicated no support at the moment).")
```
Minor nit.

File: toqito/states/tests/test_mutually_unbiased_basis.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+@pytest.mark.parametrize("dim", [2, 3, 5, 7])
+def test_mutually_unbiased_basis(dim):
+    np.testing.assert_equal(is_mutually_unbiased_basis(mutually_unbiased_basis(dim)), True)
+
+
+@pytest.mark.parametrize("dim", [4, 8, 9])
+def test_mutually_unbiased_basis_prime_power_not_prime(dim):
+    """Dimension is a prime power but not prime (this is not presently supported)."""
+    with pytest.raises(
+        ValueError, match=re.escape(f"Dimension {dim} is a prime power but not prime (more complicated no support atm).")
```
Comment (purva-thakre) [inline toqito/states/tests/test_mutually_unbiased_basis.py:20]:
```suggestion
        ValueError, match=re.escape(f"Dimension {dim} is a prime power but not prime (more complicated no support at the moment).")
```

File: toqito/states/mutually_unbiased_basis.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    else:
+        raise ValueError(f"No general construction of MUBs is known for dimension: {dim}.")
```
Comment (purva-thakre) [inline toqito/states/mutually_unbiased_basis.py:46]:
```
        raise ValueError(f"No general construction of MUBs is known for dimension: {dim}.")
```
Might want to add a link to your repo in this function's docstring or in this `ValueError` message.  

https://github.com/vprusso/qubit-upb

**Review by purva-thakre — APPROVED** [review]
LGTM except for some minor changes.

File: pyproject.toml [hunk 6]
```diff
@@ -63,6 +63,7 @@ coveralls = "^1.9"
 setuptools = ">65.5.1"
 sphinx = "6.2.1"
 sphinx_rtd_theme = "*"
+sympy = "^1.12"
```
[the remark this answers is not in the document]
Comment (vprusso) [reply pyproject.toml]: Absolutely, I'm 100% on board with that. The caret indeed I think could lead to some subtle issues depending on whether things change on the `sympy` side of things. Updated!

File: toqito/states/mutually_unbiased_basis.py (hunk 5 above)
Comment (vprusso) [reply toqito/states/mutually_unbiased_basis.py:46]: Ah, these are similarly named but different mathematical objects; this function is for mutually unbiased bases while the qubit UPB repo is unextendible product bases.

File: toqito/states/mutually_unbiased_basis.py (hunk 2 above)
Comment (vprusso) [reply toqito/states/mutually_unbiased_basis.py]: Ah, let me add that citation. Good call!

File: toqito/state_props/is_mutually_unbiased_basis.py (hunk 1 above)
Comment (vprusso) [reply toqito/state_props/is_mutually_unbiased_basis.py]: Hmm, actually, I'm not sure this check is necessary. A vector of length 1 isn't necessarily an error but is just not a MUB. I've removed this check and adapted the tests accordingly. Thanks for catching that!

**Pull request merged**